Creating new EndpointSliceProxying feature gate for kube-proxy, enabling EndpointSlice feature gate by default

robscott opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This creates a new `EndpointSliceProxying` feature gate to cover EndpointSlice consumption (kube-proxy) and allow the existing EndpointSlice feature gate to focus on EndpointSlice production only. Along with that addition, this enables the EndpointSlice feature gate by default, now only affecting the controller.

The rationale here is that it's really difficult to guarantee all EndpointSlices are created in a cluster upgrade process before kube-proxy attempts to consume them. Although masters are generally upgraded before nodes, and in most cases, the controller would have enough time to create EndpointSlices before a new node with kube-proxy spun up, there are plenty of edge cases where that might not be the case. The primary limitation on EndpointSlice creation is the API rate limit of 20QPS. In clusters with a lot of endpoints and/or with a lot of other API requests, it could be difficult to create all the EndpointSlices before a new node with kube-proxy targeting EndpointSlices spun up. 

Separating this into 2 feature gates allows for a more gradual rollout with the EndpointSlice controller being enabled by default in 1.18, and EndpointSlices for kube-proxy being enabled by default in the next release.

**Special notes for your reviewer**:
I'm not sure if the new `EndpointSliceProxying` feature gate should be alpha or beta at this point. It will not be enabled by default, but it is being split out of a feature gate that is already beta.

**Does this PR introduce a user-facing change?**:
```release-note
EndpointSlices will now be enabled by default. A new `EndpointSliceProxying` feature gate determines if kube-proxy will use EndpointSlices, this is disabled by default.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
- Enhancement issue: https://github.com/kubernetes/enhancements/issues/752


/sig network
/priority important-longterm
/cc @freehan @liggitt @thockin

robscott commented:
EndpointSlice controller has had all known bugs fixed, I think it's ready to be enabled by default now. This PR allows us to gradually enable EndpointSlices, starting with only the controller. Let me know if there's any reason not to enable this by default now.

/assign @thockin @liggitt

aojea commented:
/test pull-kubernetes-e2e-kind-ipv6

liggitt commented:
> The EndpointSlice controller has had all known bugs fixed, I think it's ready to be enabled by default now. This PR allows us to gradually enable EndpointSlices, starting with only the controller. Let me know if there's any reason not to enable this by default now.
> 
> /assign @thockin @liggitt

I haven't had a chance to review this specifically, but that matches my expectations

robscott commented:
/retest

robscott commented:
/retest

robscott commented:
/retest

thockin commented on plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/cluster-roles.yaml:
```diff
@@ -1051,6 +1051,13 @@ items:
     - create
     - patch
     - update
+  - apiGroups:
```
Why are these changes in the same commit?

robscott replied:
Good call, I've split them into two separate commits now, thanks!

robscott committed:
Creating new EndpointSliceProxying feature gate for kube-proxy

This creates a new EndpointSliceProxying feature gate to cover EndpointSlice
consumption (kube-proxy) and allow the existing EndpointSlice feature gate to
focus on EndpointSlice production only. Along with that addition, this enables
the EndpointSlice feature gate by default, now only affecting the controller.

The rationale here is that it's really difficult to guarantee all EndpointSlices
are created in a cluster upgrade process before kube-proxy attempts to consume
them. Although masters are generally upgraded before nodes, and in most cases,
the controller would have enough time to create EndpointSlices before a new node
with kube-proxy spun up, there are plenty of edge cases where that might not be
the case. The primary limitation on EndpointSlice creation is the API rate limit
of 20QPS. In clusters with a lot of endpoints and/or with a lot of other API
requests, it could be difficult to create all the EndpointSlices before a new
node with kube-proxy targeting EndpointSlices spun up.

Separating this into 2 feature gates allows for a more gradual rollout with the
EndpointSlice controller being enabled by default in 1.18, and EndpointSlices
for kube-proxy being enabled by default in the next release.

robscott committed:
Enabling EndpointSlice feature gate by default

This enables the EndpointSlice controller by default, but does not make
kube-proxy a consumer of the EndpointSlice API.

thockin reviewed (commented):
Thanks!

/lgtm
/approve
